packages/backend/server/src/plugins/copilot/utils.ts

```diff
@@ -9,3 +11,38 @@ export function readStream(
 ): Promise<Buffer> {
   return readBufferWithLimit(readable, maxSize);
 }
+
+type RequestClosedCallback = (isAborted: boolean) => void;
+type SignalReturnType = {
+  signal: AbortSignal;
+  onConnectionClosed: (cb: RequestClosedCallback) => void;
+};
+
+export function getSignal(req: Request): SignalReturnType {
+  const controller = new AbortController();
+
+  let isAborted = true;
+  let callback: ((isAborted: boolean) => void) | undefined = undefined;
+
+  const onSocketEnd = () => {
+    isAborted = true;
+  };
+  const onSocketClose = (hadError: boolean) => {
+    const aborted = hadError || isAborted;
+    if (aborted) {
+      controller.abort();
+    }
+
+    callback?.(aborted);
```

Comment:
You need to `off` the event listener first before triggering the callback to avoid the callback exception causing the `off` not to execute